### PR TITLE
Add one list account manager filter to company search.

### PR DESCRIPTION
**Pull request by elcct**
### Description of change

`POST /v3/search/company/`: new filter `one_list_group_global_account_manager` was added.

This enables filtering by computed field `one_list_group_global_account_manager` that contains an id of one list account manager. If company is a subsidiary, this value will be inherited from the parent company.

### Checklist

* [ ] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-leeloo/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/search/dict_utils.py [hunk 1]
```diff
@@ -138,11 +138,30 @@ def get_dict(obj):
...
+
+    return get_dict
+
+
 def computed_nested_id_name_dict(nested_field):
     """Creates a dictionary with id and name from a nested field."""
     return _computed_nested_dict(nested_field, id_name_dict)
 
 
+def computed_nested_contact_or_adviser_dict(nested_field):
```
Comment (reupen) [inline datahub/search/dict_utils.py]: Is this used anywhere?

File: datahub/search/company/test/test_elasticsearch.py [hunk 2]
```diff
@@ -218,6 +218,35 @@ def test_mapping(es):
                         },
                     },
                 },
+                'one_list_group_global_account_manager': {
```
Comment (reupen) [inline datahub/search/company/test/test_elasticsearch.py]: We're indexing quite a lot here, when it seems like we only need `id` to be indexed. Would be good to only index the field being filtered by, and set `index=False` on anything else.

File: datahub/search/dict_utils.py (hunk 1 above)
Comment (elcct) [reply datahub/search/dict_utils.py]: Oops, I forgot to remove. Thank you!

File: datahub/search/company/test/test_elasticsearch.py (hunk 2 above)
Comment (elcct) [reply datahub/search/company/test/test_elasticsearch.py]: That's a very good point. I replaced this field with one that only indexes the id.

**Comment by elcct** [comment]
The `one_list_account_owner` was already in the select-related fields list, but `global_headquarters__one_list_account_owner` wasn't, so I added it. Thank you!

File: datahub/search/company/test/test_entity_search_views.py [hunk 3]
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
...
+            (0, 1, 2),
+            (0, 1),
+            (0, 2),
+            (2, ),
+            (1, 0),
+        ),
+    )
+    def test_one_list_account_manager_with_global_headquarters_filter(
+        self,
+        selected_account_manager,
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Doesn't seem like `selected_account_manager` is used here.

File: datahub/search/company/test/test_entity_search_views.py [hunk 4]
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
...
+        )
+
+        assert response.status_code == status.HTTP_200_OK
+
+        search_results = {company['id'] for company in response.data['results']}
+        expected_results = {
+            str(global_headquarters.id),
+            *{str(target_company.id) for target_company in target_companies},
+        }
+        assert search_results == expected_results
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Would be good to add another assertion for the number of results as well (to make sure there are no duplicates).

File: datahub/search/company/test/test_entity_search_views.py [hunk 5]
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
...
         assert search_results == results
 
+    @pytest.mark.parametrize(
+        'selected_account_manager',
+        (
+            (0, 1, 2),
+            (0, 1),
+            (0, 2),
+            (2, ),
+            (1, 0),
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: I'm not sure on the practical difference between `(0, 1)`, `(0, 2)`, `(1, 0)`. If we want to test with 1, 2 and 3 advisers maybe we could use `random.sample(account_managers, n)` and parametrise `n`?

File: datahub/search/company/test/test_elasticsearch.py [hunk 6]
```diff
@@ -18,7 +18,6 @@ def test_mapping(es):
         CompanySearchApp.es_model.get_write_index(),
         CompanySearchApp.name,
     )
-
```
Comment (reupen) [inline datahub/search/company/test/test_elasticsearch.py]: Is this change intended?

File: datahub/search/company/test/test_entity_search_views.py [hunk 7]
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
...
+        ),
+    )
+    def test_one_list_account_manager_filter(self, selected_account_manager, es_with_signals):
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Should update these to `es_with_collector` (instead of `es_with_signal`) when we rebase.

File: datahub/search/company/test/test_entity_search_views.py (hunk 3 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: Oh. Yes, I should have removed that. Thank you!

File: datahub/search/company/test/test_entity_search_views.py (hunk 4 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: Good point! I added additional assertions now. Thank you!

File: datahub/search/company/test/test_entity_search_views.py (hunk 5 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: I changed this test to follow your suggestion. I think it is better now because it also checks the contents of `one_list_group_global_account_manager` field. Thank you!

File: datahub/search/company/test/test_elasticsearch.py (hunk 6 above)
Comment (elcct) [reply datahub/search/company/test/test_elasticsearch.py]: It is not. Fixed now. Thank you.

File: datahub/search/company/test/test_entity_search_views.py (hunk 7 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: I'll keep in mind!

**Comment by elcct** [comment]
This is a good point, I missed that! I have changed the signal to also sync subsidiaries if a company is a headquarter and added test to check that `one_list_group_global_account_manager` is populated when  `one_list_account_owner` of global hq is changed.

File: datahub/search/company/test/test_entity_search_views.py [hunk 8]
```diff
@@ -361,6 +369,132 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
...
+        search_results = {company['id'] for company in response.data['results']}
+        expected_results = {
+            str(global_headquarters.id),
+            *{str(target_company.id) for target_company in target_companies},
+        }
+        assert response.data['count'] == 3
+        assert len(response.data['results']) == 3
+        assert search_results == expected_results
+
+    def test_subsidiary_is_synced_to_es(self, es_with_signals):
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: I think we should add a specific test for the signal receiver for this (in `test_signals`) instead of testing this via the view.

File: datahub/search/company/signals.py [hunk 9]
```diff
@@ -8,10 +8,15 @@
...
-    """Sync company to the Elasticsearch."""
-    transaction.on_commit(
-        lambda: sync_object_async(CompanySearchApp, instance.pk),
-    )
+    """Sync company and subsidiaries to the Elasticsearch."""
+    def sync_es_wrapper():
+        sync_object_async(CompanySearchApp, instance.pk)
 
+        if instance.headquarter_type_id:
+            for company in instance.subsidiaries.all():
```
Comment (reupen) [inline datahub/search/company/signals.py]: It would be good to use `sync_related_objects_async` for this (similar to other search apps):

https://github.com/uktrade/data-hub-leeloo/blob/470e8ad003057b3af655fea7ea4b7d23e8d1d6d8/datahub/search/contact/signals.py#L17-L21

File: datahub/search/company/test/test_entity_search_views.py [hunk 10]
```diff
@@ -361,6 +369,132 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
         search_results = {company['name'] for company in response.data['results']}
         assert search_results == results
 
+    @pytest.mark.parametrize(
+        'n',
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Sorry, I just meant `n` as example – would be good to use something more descriptive e.g. `num_account_managers`.

File: datahub/search/company/test/test_entity_search_views.py (hunk 8 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: This is a good idea. I removed this test and created a new one in `test_signals`. Thanks!

File: datahub/search/company/signals.py (hunk 9 above)
Comment (elcct) [reply datahub/search/company/signals.py]: I did it this way first but it didn't seem to be working. Maybe I got something else wrong, because I redo it with `sync_related_objects_async` and it now works. Thank you!

File: datahub/search/company/test/test_entity_search_views.py (hunk 10 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: Haha yes I was thinking the use of `n` was odd, but perhaps didn't have enough coffee to transform it into something like `num_account_managers`. It is fixed now. Thank you!

File: changelog/company/add-one-list-group-account-manager-filter.api.md [hunk 11]
```diff
@@ -0,0 +1 @@
+`POST /v3/search/company/`: new filter `one_list_group_global_account_manager` was added.
```
Comment (currycoder) [inline changelog/company/add-one-list-group-account-manager-filter.api.md]: Should this be `POST /v4/search/company`?

Comment (elcct) [reply changelog/company/add-one-list-group-account-manager-filter.api.md]: Oh good catch! Updated now. Thank you.

File: datahub/search/company/test/test_signals.py [hunk 12]
```diff
@@ -35,3 +35,48 @@ def test_company_auto_updates_to_es(es_with_signals):
...
+    )
+    same_name = 'very_hard_to_find_company_international'
+    subsidiaries = CompanyFactory.create_batch(
+        2,
+        name=same_name,
+        global_headquarters=global_headquarters,
+    )
+    es_with_signals.indices.refresh()
+
+    result = get_basic_search_query(Company, same_name).execute()
```
Comment (reupen) [inline datahub/search/company/test/test_signals.py]: I know the other tests in this module work this way, but rather than doing a search (which seems irrelevant for testing if the signal receiver works), I think it would be better if we fetched the document directly using something like:

```python
es_with_signals.get(
    index=CompanySearchApp.es_model.get_read_alias(),
    doc_type=CompanySearchApp.name,
    id=company.pk,
)
```

(Could add a helper function to `datahub.search.test.utils` for this as well.)

Comment (elcct) [reply datahub/search/company/test/test_signals.py]: I see. I would think that `get_basic_search_query` plays a role of such helper, but I can imagine how this is not the best way to do it. I'll try to modify the test though. Thank you!

File: datahub/search/test/utils.py [hunk 13]
```diff
@@ -62,3 +62,14 @@ def db_objects_to_es_documents(db_objects, index=None):
         get_write_alias=Mock(return_value='test-write-alias'),
         get_target_index_name=Mock(return_value=f'test-index-{target_mapping_hash}'),
     )
+
+
+def get_documents_by_id(es_with_signals, app, ids):
```
Comment (reupen) [inline datahub/search/test/utils.py]: `get_documents_by_ids` might be clearer (since it takes multiple IDs).

**Comment by elcct** [comment]
I'll fix the typos whenI rebase.
